[operator-conditions] Add proposal for Operator Conditions API library

varshaprasad96 opened the pull request:
This PR adds a proposal to discuss the deliverables of the library
which would provide APIs to authors for adding/updating operator
conditions.

jmrodri reviewed (changes requested):
Are we planning on removing the operator-lib status conditions? Keeping it and adding something else to the library? Do we need to worry about backwards compatibility? I would try to answer these questions in the enhancements.

https://github.com/operator-framework/operator-lib/tree/main/status

I'm okay with what we've mentioned so far. I would just like to know more about the proposed changes to operator-lib.

varshaprasad96 commented:
@jmrodri The [status](https://github.com/operator-framework/operator-lib/tree/main/status) package present in operator-lib is very useful here. There won't be any changes made to it. The helper methods provided in there will be used in the library to work with the `conditions` present in `status.conditions` array of the CRD. I have added that in the proposal.

joelanford commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+
+The operator will be allowed to only modify the `status` field of condition CR. Operators can either add, delete or update the `status.conditions` array to include OLM-supported or custom conditions. The format and description of the fields present in the conditions can be referred [here][condition_type].
+
+The library will live in [`operator-lib`][operator_lib] repository and the definitions for Condition CRD will be present in [`operator-framwork/api`][api_repo]. The helper methods to work with [`conditions`][operator-lib-condition], which are present in operator-lib will be utilized to build APIs that will operate on the `Condition` CRD provided by OLM and perform necessary actions.
```
There's a new [upstream implementation of Conditions](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go) that I think we should transition to as part of this work.

IMO, the OLM-defined types should use `[]metav1.Condition` in their Status struct, and we should remove `operator-lib/pkg/status` in favor of the official upstream implementation.

joelanford commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+In this case, the specific operation will return an error. The returned error can be wrapped, such that users can utilize the `errors.Is` method to compare the obtained error with specific values.
+
+**NOTE:**
+Operators cannot CREATE/DELETE the `Condition CR`. OLM will scope the operators' RBAC to only GET, LIST, UPDATE the Condition CR associated with their operator.
```
We should make sure it also allows `PATCH`.

And should we document that the CRD will utilize the status subresource and operators will only be given these permissions on the status subresource (meaning no changing the spec or metadata)?

varshaprasad96 replied:
Have documented it in Risks and Mitigations

joelanford commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+
+### Risks and Mitigations
+
+- In order to restrict modifications on entire Condition CR, OLM will restrict operators' changes to only `status.conditions` field, though the entire CRD can be read.
```
Is OLM restricting the operator to the `status.conditions` field only? If so I think that would require them to include some logic in an admission webhook. I was left with the impression that the RBAC would allow the operator to change anything about the entire `status` field.

varshaprasad96 replied:
That was my mistake. You're right, operators can only edit `status` sub resource.

joelanford commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+
+This is a no-op. In this case, OLM will default to its normal behavior.
+
+2. Operator tries to get/update condition CRD when it is not present in the cluster:
+
+In this case, the specific operation will return an error. The returned error can be wrapped, such that users can utilize the `errors.Is` method to compare the obtained error with specific values.
```
This encompasses two separate cases:
1. When the CRD is not present (so the returned error would be related to that API not being found)
2. When the CRD is present, but the requested CR instance is not present (so the returned error would be related to that object not being found)

I think it would be good to spell this out to make sure we don't force users to handle the API discovery aspect of case 1.

joelanford commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+  name: foo-operator-zdgen-asdf # RandomGen Name
+  namespace: operators
+status:
+   conditions:
```
Formatting nit:
```suggestion
  conditions:
```

joelanford reviewed (commented):
Looks great overall. Just a few suggestions to add some clarity.

joelanford commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,114 @@
+
+The operator will be allowed to only modify the `status` field of condition CR. Operators can either add, delete or update the `status.conditions` array to include OLM-supported or custom conditions. The format and description of the fields present in the conditions can be referred [here][condition_type].
+
+The library will live in [`operator-lib`][operator_lib] repository and the definitions for Condition CRD will be present in [`operator-framwork/api`][api_repo].
```
I left a separate comment below, but I think we should remove our status package and direct users to migrate to https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go

There are not backwards-compat concerns here IMO since operator-lib is not 1.0.

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+
+[Operator Lifecycle Manager][olm] intends to provide operators with a channel to communicate complex states that influence its behavior while managing the operator. To do so, OLM will provide a [`Condition`][conditions_proposal] Custom Resource Definition (CRD) for the operator. Based on the [conditions][conditions] set in the CR, OLM will change its behavior accordingly.
+
+Operators need to be able to read the Condition CR created by OLM, and also have the ability to add/update necessary conditions. This library will provide APIs, using which users can read the Condition CRD present in the cluster for their operator and modify the `.status.conditions` array to add or update any specific condition.
```
```suggestion
Operators need to be able to read the Condition CR created by OLM, and also have the ability to add/update necessary conditions. This library will provide APIs that can:
* Get the singleton Condition CR managed by OLM for their operator.
* Update the CR's `.status.conditions` array.
```

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+
+### Goals:
+
+- List the deliverables of the library.
```
These deliverables should be listed as goals.

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+
+#### Story 1
+
+As an operator author, I want my operator to be able to add/modify conditions in the `Condition CR`.
```
```suggestion
As an operator author, I want my operator to be able to add/modify conditions in the `Condition` CR.
```

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+
+#### Story 2
+
+As an operator author, I would like to have the flexibility of being able to read the `Condition CR` and decide to not set any conditions, thereby opting for default OLM behavior.
```
```suggestion
As an operator author, I would like to have the flexibility of being able to read the `Condition` CR and decide to not set any conditions, thereby opting for default OLM behavior.
```

estroz replied:
Why would an operator want to get the `Condition` if it wasn't going to modify it?

varshaprasad96 replied:
Have reworded this. What I meant was, operators should have the flexibility to view the entire CR

estroz commented on enhancements/operator-conditions-lib.md:
````diff
@@ -0,0 +1,116 @@
+    lastTransitionTime: "2020-08-24T23:15:55Z"
+```
+
+The operator will be allowed to only modify the `status` field of condition CR. Operators can either add, delete or update the `status.conditions` array to include OLM-supported or custom conditions. The format and description of the fields present in the conditions can be referred [here][condition_type].
````
```suggestion
The operator will be allowed to only modify the `status` field of `Condition` CR. Operators can either add, delete or update the `status.conditions` array to include OLM-supported or custom conditions. The format and description of the fields present in the conditions can be referred [here][condition_type].
```

estroz replied:
Can you reference where "OLM-supported" conditions can be found, and a spec for writing a custom condition if one exists?

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+
+This is a no-op. In this case, OLM will default to its normal behavior.
+
+2. Operator tries to get/update condition CRD when it is not present in the cluster:
```
```suggestion
2. Operator tries to get/update the `Condition` CR when it is not present in the cluster:
```

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+In this case, the specific operation will return an error. The returned error can be wrapped, such that users can utilize the `errors.Is` method to compare the obtained error with specific values.
+
+**NOTE:**
+Operators cannot CREATE/DELETE the `Condition CR`. OLM will scope the operators' RBAC to only GET, LIST, UPDATE the Condition CR associated with their operator.
```
The `Condition` CR is a singleton for an operator, so don't mention LIST
```suggestion
Operators cannot CREATE/DELETE the `Condition CR`. OLM will scope the operators' RBAC to only GET and UPDATE the Condition CR associated with their operator.
```

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+
+### Risks and Mitigations
+
+- In order to restrict modifications on entire Condition CR, OLM will restrict operators' changes to only `status.conditions` field, though the entire CRD can be read.
```
```suggestion
- In order to restrict modifications on entire Condition CR, OLM will restrict operators' changes to only `status.conditions` field, though the entire CRD can be read.
A helpful error needs to be returned if any field other than `status.conditions` is modified,
preferably one implemented in the api library.
```

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,116 @@
+### Non-Goals:
+
+- Discuss on how OLM will process the conditions set in the CRD.
+- List the function signatures/interface types which will be present in the library.
```
Isn't this the purpose of this EP?

varshaprasad96 replied:
Added the list of functions in the EP

estroz reviewed (changes requested):
I do think this is the right place to discuss what API we want to expose to users in `operator-lib`. Otherwise this EP isn't saying much and we miss the chance to discuss how to layer this API pre-implementation.

varshaprasad96 commented:
Have addressed the review comments and made a few changes. Based on [PR](https://github.com/operator-framework/enhancements/pull/51), the name of CR provided by OLM seems to have been changed from `Condition` to `Status`.

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,144 @@
+- `Status` refers to the CR definition provided by OLM and 
+- [`metav1.Condition`][metav1-Condition] refers to specific Condition type present in `status.conditions` array.
+
+ Description             | Input(s)      | Return type(s) |  Expected output  | Remarks    |
```
Personally I'd like to see Go function signatures, which are easier to read. Ex:
````md
```go
// SetConditionStatus a metav1.Condition on an apiv1alpha1.Condition object,
// and returns an error if an error scenario is encountered (see Scenarios section).
SetConditionStatus(*apiv1alpha1.Condition, metav1.Condition) error
```
````

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,144 @@
+
+ Description             | Input(s)      | Return type(s) |  Expected output  | Remarks    |
+-------------            | ------------- | -------------| ----------------  |------------- |
+  GetStatusCRD        | Context, Runtime Object (`Status`), Object Key | `Status` struct, error | `Status` provided in input populated with values obtained from cluster |  1. Library will have helper functions to construct Object key <br/>2. Return errors as discussed in Scenarios section |
```
Can we just provide a library func that gets the `types.NamespacedName` for the `Condition` CRD instead of a `Get()` wrapper? `GetStatusCRD()` would have to either be a method on a type wrapping a `Client` or have a `Client`-type parameter, which are ugly imo.

varshaprasad96 replied:
Does this mean, the library will not provide a method to get the entire CRD from cluster? Instead we will only provide a helper to get `types.NamspacedName` and the operator would use `Get` client to fetch the CRD from cluster?

estroz replied:
Correct:
```go
// Get key.Name from env var, key.Namespace from downward API.
key, err := libcondition.GetConditionNamespacedName()
if err != nil {
	return err
}

condition := v1alpha1.Condition{}
if err = r.client.Get(ctx, key, &condition); err != nil	{
	return err
}
...
```

varshaprasad96 replied:
Reworded the EP a bit and included this.

estroz commented on enhancements/operator-conditions-lib.md:
```diff
@@ -0,0 +1,144 @@
+  SetConditionStatus     |`metav1.Condition` to add or update | error | 1. Input condition added to `Status`<br/>2. Update `metav1.Condition.status` if already exists | Return errors as discussed in Scenarios section |
+  RemoveConditionStatus  | `metav1.Condition` to delete | error |Delete condition from `status.conditions` array | Return errors as discussed in Scenarios section |
+  FindConditionStatus   | `metav1.condition.Type` | `metav1.Condition` | Return `metav1.Condition` if exists, else `nil` | 
+  IsConditionStatusTrue | `metav1.condition.Type` | bool |`True` if `metav1.condition.Status` is set | 
+  IsConditionStatusPresentAndEqual | `metav1.condition.Type`, `metav1.condition.Status`| bool |`True` if `metav1.condition.Type` is present and status is equal as provided | 
```
`FindConditionStatus()` is sufficient to check for condition truthiness/presence IMO
```suggestion
```

varshaprasad96 replied:
Would these be good to have - if I'd like to check a particular `ConditionType`, instead of getting the entire condition? Included these two from: https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go#L84-L101

estroz replied:
Couldn't you just use the meta library functions directly if you pass `apiv1alpha1.Condition.Status.Conditions`?

varshaprasad96 replied:
Yeah, but for the rest of the functions too, we are going to utilize the library functions. All of them would be wrappers around meta library functions.

varshaprasad96 replied:
Yes, that could be done too. But then, all of the functions are going to be wrappers around meta library. In the sense instead of `FindConditionStatus()` we could use the one provided in meta library and pass `apiv1alpha1.Condition.Status.conditions`

estroz reviewed (approved):
/lgtm
